### PR TITLE
Fix #22

### LINKED ISSUE #22: Jacobian for global variables
It seems that the Jacobian is not correctly evaluated when multiple individual parameters in a model error (e.g. E0 and E1) are mapped to a single global parameter (assuming E0 and E1 are to be jointly identified). In particular, in the inference problem
```
stacked_jac[:, self.latent[global_name].global_index_range()] = -J
```
stacked Jacobian has the dimension of [out, 1], whereas J has the dimension [out, 2].

### PR DESCRIPTION
**Pull request by TTitscher**
While fixing the issue, I added and refined some error messages in `bayes.latent`. The actual test case is provided in commit d9d6588 and the fix is in e631a1b.

(Edit: While this was not an urgent issue, I cannot stand having those bugs... :P )

### CODE REVIEW

File: tests/test_jacobian.py [hunk 1]
```diff
@@ -61,36 +64,94 @@ def test_scalar_prm(self):
...
+        p.latent["E"].add(me.parameter_list, "E_all")
+
+        with self.assertRaises(Exception):
+            p.jacobian([42.0])  # we have not defined a noise model yet!
+
+        noise_key = p.add_noise_model(SingleSensorNoise())
+
+        J = p.jacobian([42.0])[noise_key]
+        self.assertEqual(J.shape, (6, 1))
+        CHECK(J[:, 0], -me.x_odd - me.x_even - me.x_all)
```
Comment (joergfunger) [inline tests/test_jacobian.py:138]: I'm a little confused with the sign here. Maybe this is similar to Abbas issues with what Jacobian we are talking about. But in your model there is no data. Maybe that is part of a discussion, but I personally feel that it should be the Jacobian of the model error, not the forward model (or we provide two different methods).

File: tests/test_jacobian.py [hunk 2]
```diff
@@ -61,36 +64,94 @@ def test_scalar_prm(self):
...
+        """
+        me = OddEvenME()
+        p = VariationalBayesProblem()
+        p.add_model_error(me)
+        p.latent["E"].add(me.parameter_list, "E_odd")
+        p.latent["E"].add(me.parameter_list, "E_even")
+        noise_key = p.add_noise_model(SingleSensorNoise())
+
+        J = p.jacobian([42.0])[noise_key]
+        CHECK(J[:, 0], -me.x_odd - me.x_even)
```
Comment (joergfunger) [inline tests/test_jacobian.py:153]: same here, see above

**Review by joergfunger — APPROVED** [review]
The code itself is fine. I'm struggling with the sign of the Jacobian, in particular for the example here where there is no forward model defined, thus the argument of the return value being the Jacobian of the forward model no longer holds.

File: tests/test_jacobian.py (hunk 1 above)
Comment (TTitscher) [reply tests/test_jacobian.py:138]: This joining of Jacobians is currently only implemented in `VariationalBayesProblem`. Here, we follow the definition of Chapell where J = -dk/dtheta. Thus the minus. 

But to make that more clear, I propose to move the method *without the minus* to the base class `InferenceProblem` and add the minus only in `VariationalBayesProblem.jacobian`.

Comment (joergfunger) [reply tests/test_jacobian.py:138]: The problem I see is that even in VB we do not necessarily have to implement a forward model, just the model error. So even if its in line with the paper, the results are applicable when only implementing the model error, so my personal opinion would be to interpret the Jacobian as the derivative of the model error with respect to the parameters, and then use the minus in the update equations directly in VB - maybe by having two variables J (from the paper) = - H (d ME/d theta)

Comment (TTitscher) [reply tests/test_jacobian.py:138]: That would IMO exceed the scope here. What about just merging this PR now and handle the minus in a separate one?

**Pull request merged**